qos.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2015-2025 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+package madmin
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"gopkg.in/yaml.v2"
+)
+
+// QOSConfigVersionCurrent is the current version of the QoS configuration.
+const QOSConfigVersionCurrent = "v1"
+
+// QOSConfig represents the QoS configuration with a list of rules.
+type QOSConfig struct {
+	Version string   `yaml:"version"` // Version of the QoS configuration format
+	Rules   QOSRules `yaml:"rules"`
+}
+
+// QOSRules represents QoS rules for buckets
+type QOSRules struct {
+	Buckets []QOSBucketRule `yaml:"buckets"`
+	// Add new rules here
+}
+
+// QOSBucketRule represents a single QoS rule at the bucket level.
+type QOSBucketRule struct {
+	ID           string                `yaml:"id"`
+	Label        string                `yaml:"label,omitempty"`
+	BucketPrefix string                `yaml:"bucketPrefix"` // includes prefix if any and represents simple wildcard for bucket, prefix
```

Comment:
It seems to me that adding several buckets to a single rule would remove a lot of duplicate rules if applying to many buckets.
```suggestion
	BucketPrefixes []string                `yaml:"bucketPrefixes"` // include buckets with any of these prefixes. If none, include all.
```

We can add "UserAgent []string" or other conditions later.